src/VisualStudio/Core/Next/Remote/JsonRpcClient.cs

```diff
@@ -88,23 +93,76 @@ public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments,
         }
 
+        public void Dispose()
+        {
+            OnDisposed();
+
+            _rpc.Dispose();
+        }
+
+        protected void StartListening()
+        {
+            // due to this issue - https://github.com/dotnet/roslyn/issues/16900#issuecomment-277378950
+            // _rpc need to be explicitly started
+            _rpc.StartListening();
+        }
+
+        protected virtual void OnDisposed()
+        {
+            // do nothing
+        }
+
+        // these are for debugging purpose. once we find out root cause of the issue
+        // we will remove these.
+        private static JsonRpcDisconnectedEventArgs s_debuggingLastDisconnectReason;
+        private static string s_debuggingLastDisconnectCallstack;
+
+        private JsonRpcDisconnectedEventArgs _debuggingLastDisconnectReason;
+        private string _debuggingLastDisconnectCallstack;
+
         private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationToken)
         {
             if (cancellationToken.IsCancellationRequested)
             {
                 return true;
             }
 
-            // save extra info using NFW
-            ReportExtraInfoAsNFW(ex);
+            s_debuggingLastDisconnectReason = _debuggingLastDisconnectReason;
+            s_debuggingLastDisconnectCallstack = _debuggingLastDisconnectCallstack;
 
-            // make it to explicitly crash to get better info
-            FatalError.Report(ex);
+            // send NFW to figure out why this is happening
+            ReportExtraInfoAsNFW(ex);
 
             GC.KeepAlive(_debuggingLastDisconnectReason);
             GC.KeepAlive(_debuggingLastDisconnectCallstack);
 
-            return Contract.FailWithReturn<bool>("shouldn't be able to reach here");
+            return true;
+        }
+
+        private static bool s_reported = false;
+
+        /// <summary>
+        /// Show info bar and throw its own cancellation exception until 
+        /// we figure out this issue
+        /// https://devdiv.visualstudio.com/DevDiv/_workitems/edit/453544
+        /// </summary>
+        private void ThrowOwnCancellationToken()
+        {
+            if (CodeAnalysis.PrimaryWorkspace.Workspace != null && !s_reported)
+            {
+                // do not report it multiple times
+                s_reported = true;
+
+                // use info bar to show warning to users
+                CodeAnalysis.PrimaryWorkspace.Workspace.Services.GetService<IErrorReportingService>()?.ShowGlobalErrorInfo(
+                    ServicesVSResources.Unfortunately_a_process_used_by_Visual_Studio_has_encountered_an_unrecoverable_error_We_recommend_saving_your_work_and_then_closing_and_restarting_Visual_Studio);
+            }
+
+            // create its own cancellation token and throw it
+            var ownCancellationSource = new CancellationTokenSource();
```

Comment:
sure, I can dispose and add comment. but just for FYI. CancellationTokenSource.Dispose is only matter if any one of these are used

1. timer is used, 2. callback is registered, 3. wait handle is used. 4. linked cancellation source is created.

for this particular case, Dispose doesn't do anything.